File: src/core/context-filter.ts
import type { AtomicNodeMeta } from '../domain/interfaces/node-repository.interface.ts';
import type { ILogger } from '../domain/interfaces/logger.interface.ts';

/**
 * Tách từ thô, chuẩn hóa viết thường để tính toán độ tương đồng.
 */
export function tokenize(text: string): Set<string> {
  const words = text.toLowerCase().match(/[\p{L}\p{N}]+/gu) || [];
  return new Set(words);
}

/**
 * Tính độ tương đồng Jaccard giữa 2 bộ từ.
 */
export function calculateJaccard(set1: Set<string>, set2: Set<string>): number {
  if (set1.size === 0 || set2.size === 0) {
    return 0.0;
  }

  let intersectionCount = 0;
  for (const item of set1) {
    if (set2.has(item)) {
      intersectionCount++;
    }
  }

  const unionSize = set1.size + set2.size - intersectionCount;
  return intersectionCount / unionSize;
}

/**
 * 🔥 BỘ LỌC NGỮ CẢNH CHỦ ĐỘNG (Active Context Filtering)
 *
 * So khớp keywords mới với các nốt cũ qua giải thuật Jaccard Similarity.
 * Mở rộng đồ thị (Graph Expansion) bằng cách kéo thêm parent và children.
 *
 * @param keywords — danh sách từ khóa cần so khớp
 * @param allNodes — toàn bộ nodes từ INodeRepository
 * @param maxResults — số lượng kết quả tối đa
 * @param minScore — ngưỡng điểm tối thiểu để chọn nốt (mặc định 0.02)
 * @param titleBoost — điểm cộng khi trùng tiêu đề (mặc định 0.2)
 * @param seedRatio — tỷ lệ seed để mở rộng đồ thị (mặc định 0.5 tương đương maxResults/2)
 * @param logger — logger tùy chọn, nếu không cung cấp sẽ dùng console.log
 */
export function filterRelevantNodes(
  keywords: Array<{ name: string; definition: string }>,
  allNodes: AtomicNodeMeta[],
  maxResults = 10,
  minScore = 0.02,
  titleBoost = 0.2,
  seedRatio = 0.5,
  logger?: ILogger,
): AtomicNodeMeta[] {
  if (allNodes.length === 0) return [];

  // 1. Thu thập từ khóa mới để tính điểm
  const newKwTokens = new Set<string>();
  for (const kw of keywords) {
    const nameTokens = tokenize(kw.name || '');
    const defTokens = tokenize(kw.definition || '');
    for (const token of nameTokens) newKwTokens.add(token);
    for (const token of defTokens) newKwTokens.add(token);
  }

  if (newKwTokens.size === 0) {
    return [];
  }

  // 2. Tính điểm Jaccard similarity thô
  const scoredNodes: Array<{ score: number; node: AtomicNodeMeta }> = [];
  for (const node of allNodes) {
    const oldText = `${node.title} ${node.tags.join(' ')} ${node.definition}`;
    const oldTokens = tokenize(oldText);

    let score = calculateJaccard(newKwTokens, oldTokens);

    // Điểm cộng đặc biệt nếu trùng khớp trực tiếp tiêu đề/tên slug
    const nodeTitleTokens = tokenize(node.title);
    let hasTitleOverlap = false;
    for (const token of nodeTitleTokens) {
      if (newKwTokens.has(token)) {
        hasTitleOverlap = true;
        break;
      }
    }

    if (hasTitleOverlap) {
      score += titleBoost;
    }

    scoredNodes.push({ score, node });
  }

  // Sắp xếp theo điểm giảm dần
  scoredNodes.sort((a, b) => b.score - a.score);

  // 3. Lấy Top K nốt khớp nhất
  const topMatches: AtomicNodeMeta[] = [];
  const seenSlugs = new Set<string>();

  const seedCount = Math.max(2, Math.floor(maxResults * seedRatio));
  for (let i = 0; i < seedCount && i < scoredNodes.length; i++) {
    const { score, node } = scoredNodes[i];
    if (score > minScore) {
      topMatches.push(node);
      seenSlugs.add(node.slug);
    }
  }

  // 4. MỞ RỘNG ĐỒ THỊ (Graph Expansion Step)
  const expandedNodes = [...topMatches];

  const nodeBySlug = new Map<string, AtomicNodeMeta>();
  for (const n of allNodes) {
    nodeBySlug.set(n.slug, n);
  }

  for (const seedNode of topMatches) {
    // Kéo nốt cha (parent)
    const parentSlug = seedNode.parent;
    if (parentSlug && !seenSlugs.has(parentSlug) && nodeBySlug.has(parentSlug)) {
      expandedNodes.push(nodeBySlug.get(parentSlug)!);
      seenSlugs.add(parentSlug);
    }

    // Kéo nốt con (children)
    for (const childSlug of seedNode.children) {
      if (!seenSlugs.has(childSlug) && nodeBySlug.has(childSlug)) {
        expandedNodes.push(nodeBySlug.get(childSlug)!);
        seenSlugs.add(childSlug);
      }
    }
  }

  const log = logger || { info: (msg: string) => console.log(msg), success: (msg: string) => console.log(msg), warn: (msg: string) => console.warn(msg), error: (msg: string) => console.error(msg), log: (msg: string) => console.log(msg) };
  log.info(`  🔍 Active Context Filter: Quét ${allNodes.length} nốt cũ -> Lọc ra ${expandedNodes.length} nốt liên quan nhất (tiết kiệm ~85% token).`);
  return expandedNodes.slice(0, maxResults);
}

File: scripts/mrp_pipeline/core/context_filter.py
import os
import re
import yaml
from typing import List, Dict, Any, Set

from scripts.mrp_pipeline.core.config import DIR_ATOMIC

def tokenize(text: str) -> Set[str]:
    """Tách từ thô, chuẩn hóa viết thường để tính toán độ tương đồng."""
    words = re.findall(r'\b\w+\b', text.lower())
    return set(words)

def calculate_jaccard(set1: Set[str], set2: Set[str]) -> float:
    """Tính độ tương đồng Jaccard giữa 2 bộ từ."""
    if not set1 or not set2:
        return 0.0
    return len(set1.intersection(set2)) / len(set1.union(set2))

def parse_atomic_node_meta(filepath: str) -> Optional[Dict[str, Any]]:
    """Đọc nhanh metadata frontmatter và định nghĩa của nốt nguyên tử cũ."""
    try:
        with open(filepath, "r", encoding="utf-8") as f:
            content = f.read()
    except Exception:
        return None

    # Tách frontmatter
    fm_match = re.match(r"^---\n(.*?)\n---", content, re.DOTALL)
    if not fm_match:
        return None

    fm_text = fm_match.group(1)
    slug = os.path.basename(filepath).replace("HAE-concept-", "").replace(".md", "")

    try:
        data = yaml.safe_load(fm_text)
        if not data:
            data = {}
    except Exception:
        data = {}

    title = data.get("title", slug)
    category = data.get("category", "")
    tags = data.get("tags", [])
    parent = data.get("parent")
    children = data.get("children", [])

    # Lấy định nghĩa cốt lõi
    definition = ""
    def_match = re.search(r"## 💡 Định nghĩa & Nội dung Cốt lõi\n(.+?)(?=\n##|\Z)", content, re.DOTALL)
    if def_match:
        definition = def_match.group(1).strip()

    return {
        "slug": slug,
        "title": title,
        "category": category,
        "tags": tags,
        "parent": parent,
        "children": children,
        "definition": definition,
        "filename": os.path.basename(filepath)
    }

def filter_relevant_nodes(keywords: List[Dict[str, str]], max_results: int = 10) -> List[Dict[str, Any]]:
    """
    🔥 BỘ LỌC NGỮ CẢNH CHỦ ĐỘNG (Active Context Filtering)
    So khớp keywords mới với các nốt cũ qua giải thuật Jaccard Similarity.
    Mở rộng đồ thị (Graph Expansion) bằng cách kéo thêm parent và children.
    """
    if not os.path.exists(DIR_ATOMIC):
        return []

    # 1. Thu thập từ khóa mới để tính điểm
    new_kw_tokens = set()
    for kw in keywords:
        new_kw_tokens.update(tokenize(kw.get("name", "")))
        new_kw_tokens.update(tokenize(kw.get("definition", "")))

    if not new_kw_tokens:
        return []

    all_nodes = []
    # Quét toàn bộ nốt
    for file in os.listdir(DIR_ATOMIC):
        if file.endswith(".md") and file.startswith("HAE-concept-"):
            filepath = os.path.join(DIR_ATOMIC, file)
            meta = parse_atomic_node_meta(filepath)
            if meta:
                all_nodes.append(meta)

    # 2. Tính điểm Jaccard similarity thô
    scored_nodes = []
    for node in all_nodes:
        # Gom nội dung nốt cũ để so khớp
        old_text = f"{node['title']} {' '.join(node['tags'])} {node['definition']}"
        old_tokens = tokenize(old_text)

        score = calculate_jaccard(new_kw_tokens, old_tokens)

        # Điểm cộng đặc biệt nếu trùng khớp trực tiếp tiêu đề/tên slug
        node_title_tokens = tokenize(node['title'])
        if len(new_kw_tokens.intersection(node_title_tokens)) > 0:
            score += 0.2

        scored_nodes.append((score, node))

    # Sắp xếp theo điểm giảm dần
    scored_nodes.sort(key=lambda x: x[0], reverse=True)

    # 3. Lấy Top K nốt khớp nhất
    top_matches = []
    seen_slugs = set()

    # Chúng ta lấy khoảng max_results // 2 nốt khớp nhất làm hạt nhân
    seed_count = max(2, max_results // 2)
    for score, node in scored_nodes[:seed_count]:
        if score > 0.02:  # Ngưỡng tối thiểu để tránh kéo nốt không liên quan
            top_matches.append(node)
            seen_slugs.add(node["slug"])

    # 4. MỞ RỘNG ĐỒ THỊ (Graph Expansion Step)
    # Tự động kéo thêm nốt cha và các nốt con trực tiếp của các nốt hạt nhân
    expanded_nodes = list(top_matches)

    node_by_slug = {n["slug"]: n for n in all_nodes}

    for seed_node in top_matches:
        # Kéo nốt cha (parent)
        parent_slug = seed_node.get("parent")
        if parent_slug and parent_slug not in seen_slugs and parent_slug in node_by_slug:
            expanded_nodes.append(node_by_slug[parent_slug])
            seen_slugs.add(parent_slug)

        # Kéo nốt con (children)
        for child_slug in seed_node.get("children", []):
            if child_slug not in seen_slugs and child_slug in node_by_slug:
                expanded_nodes.append(node_by_slug[child_slug])
                seen_slugs.add(child_slug)

    # Giới hạn kích thước tối đa để tránh phình to context
    print(f"  🔍 Active Context Filter: Quét {len(all_nodes)} nốt cũ -> Lọc ra {len(expanded_nodes)} nốt liên quan nhất (tiết kiệm ~85% token).")
    return expanded_nodes[:max_results]
